src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,92 @@
+        /// Note that DiagnosticHandler is the latest in the pipeline </param>
+        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler)
+        {
+        }
+
+        internal static bool IsEnabled()
+        {
+            return s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.ResponseWriteName);
+        }
+
+        protected internal override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            //HttpClientHandler is responsible to call DelegatingHandler.IsEnabled() before forwarding request here.
+            //This code will not be reached if ResponseWriteName is not enabled, unless consumer unsubscribes
+            //from DiagnosticListener right after the check. So some requests happening right after subscription starts
+            //might not be instrumented. Similarly, when consumer unsubscribes, extra requests might be instumented
+
+            Guid loggingRequestId = Guid.NewGuid();
+            if (s_diagnosticListener.IsEnabled(DiagnosticsHandlerLoggingStrings.RequestWriteName))
+            {
+                LogHttpRequest(request,  loggingRequestId);
+            }
+            Task<HttpResponseMessage> responseTask = base.SendAsync(request, cancellationToken);
+
+            LogHttpResponse(responseTask, loggingRequestId);
+            return responseTask;
+        }
+
+        #region private
+
+        private static readonly DiagnosticListener s_diagnosticListener =
+            new DiagnosticListener(DiagnosticsHandlerLoggingStrings.DiagnosticListenerName);
+
+        private static void LogHttpRequest(HttpRequestMessage request, Guid loggingRequestId)
+        {
+            long timestamp = Stopwatch.GetTimestamp();
+
+            s_diagnosticListener.Write(
+                DiagnosticsHandlerLoggingStrings.RequestWriteName,
+                new
+                {
+                    Request = request,
+                    LoggingRequestId = loggingRequestId,
+                    Timestamp = timestamp
+                }
+            );
+        }
+
+        private static void LogHttpResponse(Task<HttpResponseMessage> responseTask, Guid loggingRequestId)
+        {
+            responseTask.ContinueWith(
+                (t, s) =>
+                {
+                    long timestamp = Stopwatch.GetTimestamp();
+
+                    s_diagnosticListener.Write(
+                        DiagnosticsHandlerLoggingStrings.ResponseWriteName,
+                        new
+                        {
+                            Response = t.Status == TaskStatus.RanToCompletion ? t.Result : null,
+                            LoggingRequestId = (Guid)s,
+                            TimeStamp = timestamp,
+                            Exception = t.Exception,
+                            RequestTaskStatus = t.Status
+                        }
```

Comment:
It is a fair question, and currently I don't have a strong opinion.   In such situation I typically take my cue from the code.    What convention makes the code cleaner (both for creation and consumption).    Certainly from a start-stop perspective, it is GOOD if every start has a stop and the success/failure of the activity does not change the event used as the stop.   

On the other hand, often errors are much more interesting than success cases, so it seem very GOOD if errors fired their own events (and thus you can easily filter to just them).    However this might be taken care of simply by having an event for ANY exceptions (which cover errors regardless of component), so you don't need special error events in most cases.  

Whatever we do, ideally we would do it uniformly (we don't have different conventions for different components.  

My inclination, however is to ALWAYS have stop events (independent of error, but with an indication in the payload) like we have above, BUT we may ADDTIONALLY want an event that fires on an error (before the stop event fires, because it makes it clear that the error event is nested inside the activity).

But these are just inclinations.     If this proposal has bad properties that we have not yet considered, we should certainly continue the discussion...